cpp/include/raft/spatial/knn/detail/knn_brute_force_faiss.cuh

```diff
@@ -261,15 +263,18 @@ void brute_force_knn_impl(std::vector<float *> &input,
     out_I = all_I.data();
   }
 
-  // Sync user stream only if using other streams to parallelize query
-  if (n_int_streams > 0) CUDA_CHECK(cudaStreamSynchronize(userStream));
+  // Make other streams from pool wait on main stream
+  if (handle.is_stream_pool_initialized()) {
+    handle.wait_stream_pool_on_stream();
+  }
 
   for (size_t i = 0; i < input.size(); i++) {
     float *out_d_ptr = out_D + (i * k * n);
     IdxType *out_i_ptr = out_I + (i * k * n);
 
-    cudaStream_t stream =
-      raft::select_stream(userStream, internalStreams, n_int_streams, i);
+    auto stream = handle.is_stream_pool_initialized()
+                    ? handle.get_stream_from_stream_pool()
+                    : handle.get_stream();
```

Comment:
The user wants to select another stream from the pool, or fall back to the user stream. This pattern also appears [here](https://github.com/rapidsai/cuml/blob/067344041b1563b19301e2e69240a56605a67997/cpp/src_prims/linalg/lstsq.cuh#L254-L260), and I think it is a generally useful pattern.

I suggest that we add a small helper to the handle that does this (not sure about the right naming):
```c++
rmm::cuda_stream_view raft::handle::get_next_stream() {
  return handle.is_stream_pool_initialized()
                    ? handle.get_stream_from_stream_pool()
                    : handle.get_stream();
}
```

Reply:
I like the idea! Maybe `get_any_worker_stream`? cc @cjnolet for ideas

Reply:
With my limited knowledge of RMM's stream pool API, the better name depend on the expected behavior, IMO. For example, if get_stream_from_stream_pool` is acting like a thread pool and returning any available stream, then I think `get_any_worker_stream` might be the better name because it doesn't imply anything about order. if `get_stream_from_stream_pool` always returns the next stream in the pool, I would be inclined to prefer `get_next_stream()`. 

Though neither of these seem to imply anything about the fact that the default stream on the handle is used if there's no stream is configured, which makes me think that maybe a more concise name could be something like `get_usable_stream()` or `get_next_usable_stream()`

Reply:
`get_stream_from_stream_pool` always returns the next stream in the pool. In that case, I will use `get_next_usable_stream` and clarify the expected behavior in the docs